drivers/analog-io/max2201x/iio_max2201x.c

```diff
@@ -0,0 +1,919 @@
+				     const struct iio_ch_info *channel,
+				     intptr_t priv)
+{
+	enum max2201x_channel ch = channel->address;
+	struct max2201x_iio_desc *desc = dev;
+	int16_t val;
+
+	iio_parse_value(buf, IIO_VAL_INT, (int32_t *)&val, NULL);
+
+	return max2201x_offset_calib(desc->max2201x_desc, ch, val);
+}
+
+static int max2201x_iio_write_gain(void *dev, char *buf, uint32_t len,
+				   const struct iio_ch_info *channel,
+				   intptr_t priv)
+{
+	enum max2201x_channel ch = channel->address;
+	struct max2201x_iio_desc *desc = dev;
+	uint16_t val;
+
+	iio_parse_value(buf, IIO_VAL_INT, (int32_t *)&val, NULL);
+
+	return max2201x_gain_calib(desc->max2201x_desc, ch, val);
+}
+
+static int max2201x_iio_write_slew_rate(void *dev, char *buf, uint32_t len,
+					const struct iio_ch_info *channel,
+					intptr_t priv)
+{
+	enum max2201x_channel ch = channel->address;
+	struct max2201x_iio_desc *desc = dev;
+	uint32_t val, i;
+	enum max2201x_sr_step_size step_size;
+	enum max2201x_sr_update_rate update_rate;
+	bool enable, hart;
+	int ret;
+
+	ret = max2201x_reg_read(desc->max2201x_desc,
+				ch ? MAX2201X_AO_SLEW_RATE_CH2 : MAX2201X_AO_SLEW_RATE_CH1, &val);
+	if (ret)
+		return ret;
+
+	step_size = no_os_field_get(MAX2201X_SR_STEP_SIZE_MASK, val);
+	update_rate = no_os_field_get(MAX2201X_SR_UPDATE_RATE_MASK, val);
+	enable = no_os_field_get(MAX2201X_SR_EN_MASK, val);
+	hart = no_os_field_get(MAX2201X_SR_SEL_MASK, val);
+
+	switch (priv) {
+	case MAX2201X_IIO_SLEW_RATE:
+		for (i = 0; i < NO_OS_ARRAY_SIZE(max2201x_enable_avail); i++)
+			if (!strcmp(buf, max2201x_enable_avail[i]))
+				break;
+
+		if (i == NO_OS_ARRAY_SIZE(max2201x_enable_avail))
+			return -EINVAL;
+
+		enable = i;
+		break;
+	case MAX2201X_IIO_RANGE:
+		for (i = 0; i < NO_OS_ARRAY_SIZE(max2201x_range_avail); i++)
+			if (!strcmp(buf, max2201x_range_avail[i]))
+				break;
+
+		if (i == NO_OS_ARRAY_SIZE(max2201x_range_avail))
+			return -EINVAL;
+
+		hart = i;
+		break;
+	case MAX2201X_IIO_STEP_SIZE:
+		for (i = 0; i < NO_OS_ARRAY_SIZE(max2201x_step_size_avail); i++)
+			if (!strcmp(buf, max2201x_step_size_avail[i]))
+				break;
+
+		if (i == NO_OS_ARRAY_SIZE(max2201x_step_size_avail))
+			return -EINVAL;
+
+		step_size = i;
+		break;
+	case MAX2201X_IIO_UPDATE_RATE:
+		for (i = 0; i < NO_OS_ARRAY_SIZE(max2201x_update_rate_avail); i++)
+			if (!strcmp(buf, max2201x_update_rate_avail[i]))
+				break;
+
+		if (i == NO_OS_ARRAY_SIZE(max2201x_update_rate_avail))
+			return -EINVAL;
+
+		update_rate = i;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return max2201x_slew_rate(desc->max2201x_desc, ch, step_size, update_rate,
+				  enable, hart);
+}
+
+static int max2201x_iio_read_step_size_available(void *dev, char *buf,
+		uint32_t len, const struct iio_ch_info *channel, intptr_t priv)
+{
+	uint32_t length = 0;
+	uint32_t i;
+
+	for (i = 0; i < NO_OS_ARRAY_SIZE(max2201x_step_size_avail); i++)
+		length += sprintf(buf + length, "%s ", max2201x_step_size_avail[i]);
```

Comment:
this looks like it is being repeated a few times, maybe also this can be replaced with a function which takes as parameter the array, its size, the buffer and returns the length